pkg/services/ngalert/api/api_ruler.go

```diff
@@ -83,24 +88,23 @@ func (srv RulerSrv) RouteDeleteAlertRules(c *models.ReqContext) response.Respons
 			return nil
 		}
 
-		canDelete = make([]string, 0, len(q.Result))
-		for _, rule := range q.Result {
-			if authorizeDatasourceAccessForRule(rule, hasAccess) {
-				canDelete = append(canDelete, rule.UID)
-				continue
-			}
-			cannotDelete = append(cannotDelete, rule.UID)
+		deletableRulesFGAC, err := srv.accessibleRulesFGAC(q.Result, hasAccess, logger)
+		if err != nil {
+			return err
 		}
 
-		if len(canDelete) == 0 {
-			return fmt.Errorf("%w to delete rules because user is not authorized to access data sources used by the rules", ErrAuthorization)
+		deletableRulesProvisioning, err := srv.accessibleRulesProvisioning(q.Result, provenances, logger)
+		if err != nil {
+			return err
 		}
 
-		if len(cannotDelete) > 0 {
-			logger.Info("user cannot delete one or many alert rules because it does not have access to data sources. Those rules will be skipped", "expected", len(q.Result), "authorized", len(canDelete), "unauthorized", cannotDelete)
+		for uid := range deletableRulesFGAC {
+			if _, exists := deletableRulesProvisioning[uid]; exists {
+				deletableRules = append(deletableRules, uid)
+			}
```

Comment:
Rather than getting the list of FGAC-deletable and Provisioning-deletable rules (making many passes over and copies of data), would it be cleaner to structure the checks like "filters" and loop over once, sort of like a small rule engine?
```
for rule := q.Result {
    if ok, err := isDeletableFGAC(rule, logger); !ok {  //  Emit logs from inside rule.
        if err != nil {
            return err
        }
        continue
    }
    if ok, err := isDeletableProvisioning(rule, logger); !ok {
        if err != nil {
            return err
        }
        continue
    }
    deletableRules = append(deletableRules, uid)
}
```
I think this results in a little more natural error path behavior (we don't leak provisioning metadata  on non-FGAC-accessible rules), is a little easier to extend, and avoids some copies

Reply:
This approach would not let you tell the user why exactly the group could not be deleted (either RBAC or provenance).

Reply:
As alternative 
```go
	var canDelete []*ngmodels.AlertRule
	var cannotDelete []string

	partition := func(alerts []*ngmodels.AlertRule, predicate func(rule *ngmodels.AlertRule) bool) ([]*ngmodels.AlertRule, []string) {
		positive, negative := make([]*ngmodels.AlertRule, 0, len(alerts)), make([]string, 0, len(alerts))
		for _, rule := range alerts {
			if predicate(rule) {
				positive = append(positive, rule)
				continue
			}
			negative = append(negative, rule.UID)
		}
		return positive, negative
	}

   .... 
   
   		canDelete, cannotDelete = partition(q.Result, func(rule *ngmodels.AlertRule) bool {
			return authorizeDatasourceAccessForRule(rule, hasAccess)
		})
		if len(canDelete) == 0 {
			return fmt.Errorf("%w to delete rules because user is not authorized to access data sources used by the rules", ErrAuthorization)
		}
		if len(cannotDelete) > 0 {
			logger.Info("user cannot delete one or many alert rules because it does not have access to data sources. Those rules will be skipped", "expected", len(q.Result), "authorized", len(canDelete), "unauthorized", cannotDelete)
		}

		canDelete, cannotDelete = partition(canDelete, func(rule *ngmodels.AlertRule) bool {
			provenance, exists := provenances[rule.UID]
			return (exists && provenance == ngmodels.ProvenanceNone) || !exists
		})

		if len(canDelete) == 0 {
			return fmt.Errorf("all rules have been provisioned and cannot be deleted through this api")
		}

		if len(cannotDelete) > 0 {
			logger.Info("user cannot delete one or many alert rules because it does has a provenance set. Those rules will be skipped", "expected", len(rules), "provenance_none", len(canDelete), "provenance_set", cannotDelete)
		}
```

Reply:
partition function is a pure one and could be public

Reply:
I really like this, will change it.